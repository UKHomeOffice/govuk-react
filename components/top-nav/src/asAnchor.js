import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import { WHITE, YELLOW } from 'govuk-colours';

const asAnchor = (AnchorType) => {
  const Anchor = props => (
    <AnchorType {...props}>{props.children}</AnchorType>
  );

  Anchor.propTypes = {
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
      .isRequired,
  };

  const StyledHoc = glamorous(Anchor)({
    color: WHITE, // TODO: active state, LIGHT_BLUE
    textDecoration: 'none',
    borderBottom: '1px solid transparent',
    fontWeight: 700,
    ':hover': {
      borderBottomColor: WHITE, // TODO: active state, LIGHT_BLUE
    },
    ':focus': {
      backgroundColor: YELLOW,
      outline: `3px solid ${YELLOW}`,
    },
    ':visited': {
      color: WHITE, // TODO: active state, LIGHT_BLUE
      borderBottomColor: WHITE, // TODO: active state, LIGHT_BLUE
    },
  });

  return StyledHoc;
};

export default asAnchor;
