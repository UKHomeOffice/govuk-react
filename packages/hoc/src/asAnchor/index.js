import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import { BLUE, PURPLE, YELLOW } from 'govuk-colours';

const asAnchor = (AnchorType) => {
  const Anchor = props => (
    <AnchorType {...props}>{props.children}</AnchorType>
  );

  Anchor.propTypes = {
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
      .isRequired,
    onClick: PropTypes.func,
    disabled: PropTypes.bool,
  };

  Anchor.defaultProps = {
    onClick: undefined,
    disabled: undefined,
  };

  const StyledHoc = glamorous(Anchor)({
    color: BLUE,
    padding: '3px',
    margin: '-3px',
    outlineColor: 'transparent',
    display: 'inline-block',
    textDecoration: 'underline',
    ':focus': {
      backgroundColor: YELLOW,
      outline: `3px solid ${YELLOW}`,
    },
    ':visited': {
      color: PURPLE,
    },
  });

  return StyledHoc;
};

export default asAnchor;
