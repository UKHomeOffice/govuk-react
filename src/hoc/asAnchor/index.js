import React from "react";
import PropTypes from "prop-types";
import glamorous from "glamorous";
import { BLUE, PURPLE, YELLOW } from "govuk-colours";

const AnchorWrapper = glamorous.span({
  "> a": {
    color: BLUE,
    padding: "3px",
    margin: "-3px",
    outlineColor: "transparent",
    display: "inline-block",
    textDecoration: "underline",
    ":focus": {
      backgroundColor: YELLOW,
      outline: `3px solid ${YELLOW}`
    },
    ":visited": {
      color: PURPLE
    }
  }
});

const asAnchor = AnchorType => {
  const Anchor = props => (
    <AnchorWrapper>
      <AnchorType {...props}>{props.children}</AnchorType>
    </AnchorWrapper>
  );

  Anchor.propTypes = {
    children: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
      .isRequired,
    onClick: PropTypes.func,
    disabled: PropTypes.bool
  };

  Anchor.defaultProps = {
    onClick: undefined,
    disabled: undefined
  };

  return Anchor;
};

export default asAnchor;
