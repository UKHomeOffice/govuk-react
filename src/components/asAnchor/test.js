import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import asAnchor from "./";

const AnchorTag = asAnchor("a");

const wrapper = <AnchorTag href="https://example.com">Example</AnchorTag>;

describe(asAnchor, () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(wrapper, div);
  });

  it("matches wrapper snapshot", () => {
    expect(shallow(wrapper)).toMatchSnapshot(`wrapper mount`);
  });
});
