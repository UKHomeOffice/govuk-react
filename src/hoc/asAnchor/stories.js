import React from 'react';
import { storiesOf } from '@storybook/react';
import { BrowserRouter, Link } from 'react-router-dom';

import asAnchor from '.';

const AnchorTag = asAnchor('a');
const AnchorLink = asAnchor(Link);

storiesOf('asAnchor', module).add('asAnchor passing in anchor', () => (
  <AnchorTag href="https://cats.org">AsAnchor anchor example</AnchorTag>
));

storiesOf('asAnchor', module).add(
  'asAnchor passing in anchor target=blank',
  () => (
    <AnchorTag href="https://cats.org" target="_blank">
      AsAnchor anchor example
    </AnchorTag>
  ),
);

storiesOf('asAnchor', module).add(
  'asAnchor passing in React Router Link',
  () => (
    <BrowserRouter>
      <AnchorLink to="https://dogs.org">AsAnchor Link example</AnchorLink>
    </BrowserRouter>
  ),
);
