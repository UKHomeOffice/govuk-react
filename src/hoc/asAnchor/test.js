import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import asAnchor from "./";

const AnchorTag = asAnchor("a");

const wrapper = <AnchorTag href="https://example.com">Example</AnchorTag>;

describe(asAnchor, () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(wrapper, div);
  });

  it("returns a component", () => {
    expect(shallow(wrapper).html().toBeTruthy);
  });

  it("matches wrapper snapshot", () => {
    expect(shallow(wrapper)).toMatchSnapshot(`wrapper mount`);
  });
});
