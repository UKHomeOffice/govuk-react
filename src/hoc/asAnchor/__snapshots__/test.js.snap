// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asAnchor matches wrapper snapshot: wrapper mount 1`] = `
<glamorous(span)>
  <a
    href="https://example.com"
  >
    Example
  </a>
</glamorous(span)>
`;
